app/controllers/registrant/contacts_controller.rb

```diff
@@ -25,4 +47,79 @@ def domain_ids
       BusinessRegistryCache.fetch_by_ident_and_cc(ident, ident_cc).associated_domain_ids
     end
   end
-end
+
+  def domain
+    current_user_domains.find(params[:domain_id])
+  end
+
+  def current_user_domains
+    ident_cc, ident = current_registrant_user.registrant_ident.split '-'
+    begin
+      BusinessRegistryCache.fetch_associated_domains ident, ident_cc
+    rescue Soap::Arireg::NotAvailableError => error
+      flash[:notice] = I18n.t(error.json[:message])
+      Rails.logger.fatal("[EXCEPTION] #{error.to_s}")
+      current_registrant_user.domains
+    end
+  end
+
+  def contact_params
+    permitted = %i[
+      name
+      email
+      phone
+    ]
+
+    permitted << :fax if fax_enabled?
+    permitted += %i[street zip city state country_code] if Contact.address_processing?
+    params.require(:contact).permit(*permitted)
+  end
+
+  def access_token
+    uri = URI.parse("#{ENV['registrant_api_url']}/v1/registrant/auth/eid")
+    request = Net::HTTP::Post.new(uri)
+    request.form_data = { ident: current_registrant_user.registrant_ident.split('-').second,
+                          first_name: 'test',
+                          last_name: 'test' }
+
+    response = Net::HTTP.start(uri.hostname, uri.port) do |http|
+      http.request(request)
+    end
+
+    json_doc = JSON.parse(response.body, symbolize_names: true)
+    json_doc[:access_token]
+  end
+
+  def fax_enabled?
+    ENV['fax_enabled'] == 'true'
+  end
+
+  def contact_update_api_params
+    params = contact_params
+    params = normalize_address_attributes_for_api(params) if Contact.address_processing?
+    params
+  end
+
+  def normalize_address_attributes_for_api(params)
+    normalized = params
+
+    Contact.address_attribute_names.each do |attr|
+      attr = attr.to_sym
+      normalized["address[#{attr}]"] = params[attr]
+      normalized.delete(attr)
+    end
+
+    normalized
+  end
+
+  def update_contact_via_api(uuid)
```

Comment:
- https://github.com/puma/puma/issues/1433#issuecomment-417258131
- What are the other options? Or you mean that HTTP 500 is not handled in API?
- Currently it still writes to DB anyway. If we have a plan to make Registrant Portal to communicate exclusively via API, then of cource it's worth getting contact details via it, and not from DB (if this is what you mean).

Reply:
What I mean is that Registrant portal has direct connection to database that we use for other actions. There's no need to make this particular action call the API, which presents a problem because suddenly you have an intermediary that might fail (and this implementation exposes that failure in a very transpararent way).  Based on what I know, we have two options:

- Remove this from the portal completely, before the end of the year we will have a registrant portal as a separate application that communicates exclusively via REST API.
- Make it a plain Rails controller action because it already has access to the database. It will still be rendered obsolete by the end of the year.

I personally think if second option is too much work, first one is a lot safer.

Reply:
> There's no need to make this particular action call the API

Before there was. @vohmar expected that the current version of Registrant Portal already communicates exclusively via API, but for some unknown reason it doesn't. I even didn't consider other options, assumed that it should use API by default.

If I remake it so that it doesn't use API, what was the point developing at all?

Reply:
> Before there was. @vohmar expected that the current version of Registrant Portal already communicates exclusively via API, but for some unknown reason it doesn't. I even didn't consider other options, assumed that it should use API by default.

Contact update functionality was required around the time of GDPR enforcement.
Current version of registrant portal never communicated via REST API, because REST API is not even live yet.

 It also is about to be retired if favour of an application that is ordered to fit in with the new branding.

Reply:
It's up to @vohmar to decide what to do next with this issue. Given it works, and the fact that currently Registrant Portal is used quite rarely (correct me if I am wrong), and the fact that the entire portal in its current form will be discarded anyway, I would leave it as it.